blog/_src/posts/2017-05-12-artifacts-for-semantics.md

```diff
@@ -0,0 +1,155 @@
+resulted in something that is more accessible to readers, and thus worth
+considering.
+
+This post will summarize the different parts to make it easier for others to
+repeat this process. We think it was a total success, and have gotten feedback
+that it makes understanding the (somewhat complex) language from the paper much
+easier. We argue that building such interpreters / typecheckers is easy enough
+that all papers should do this. Further, while our interpreter / typechecker is
+completely unverified, since we wrote it in OCaml, this approach should work
+equally well for semantics verified in Coq and then extracted to OCaml.
+
+------
+
+The paper in question,
+[FunTAL: Reasonably Mixing a Functional Language with Assembly](https://dbp.io/pubs/2017/funtal.pdf) (to
+appear in PLDI17), presents a multi-language that incorporates a typed assembly
+language and a simple functional language where each can be embedded within the
+other. The paper then develops a logical relation that can be used to reason
+about equivalence of such mixed programs. For example in the paper we show an
+imperative register-based factorial and a functional factorial equivalent.
+
+Both the static and dynamic semantics are relatively complex. The typed assembly
+has registers (which store word-sized values), a heap (which stores code-blocks
+and tuples), and a stack (not a call-stack, simply a place where word-sized
+values can be pushed and popped). Code-blocks have pre-conditions on the state
+of the registers and the stack, and allow the tail of the stack to be abstracted
+over polymorphically. This allows values to be protected on the stack before
+jumping to blocks that otherwise could change them. This is used, along with a
+novel notion of **return markers**, to ensure well-bracketing in the control
+flow of the typed assembly. The return markers indicate the location that points
+to the block that will eventually be returned to (assuming it doesn't loop
+infinitely). At the top level, the return marker `end` indicates that, assuming
+it does not loop, eventually the program will stop, rather than returning
+somewhere else. 
+
+Understanding the dynamic semantics requires tracking how values flow through
+the registers, the heap, and the stack, and rather than a call-stack, the user
+has to track the control flow through the statically-enforced return markers.
+This allows a good deal of low-level control-flow while still ensuring that
+calls will eventually return to the right place. This well-bracketing is vital
+to be able to reason about "components" that eventually return a value of a
+particular type, a necessity when embedding these components in a typed
+high-level program! However, it does mean that understanding it from a few
+typing rules alone is a tall order. Our functional language is more standard,
+though we use iso-recursive types to allow recursion, which can easily trip up
```

Comment:
You could write `(iso)-recursive` to emphasize that the iso/equi distinction is not emphasized.